Add support for DNS01 challenges

XaF opened the pull request:
Created another Dockerfile, pretty much a copy of the normal one, but with dns-cloudflare support so the DNS01 challenge type can be used, and support for wildcard certificates is also possible.

XaF commented:
> This is an interesting addition to the feature set of this image, and I think that if we are to support DNS challenges we should include it in the current Debian and Alpine images, so we can probably begin by adding this change directly to them.

I hesitated at first, I didn't know if you wanted to keep the images as small as possible (hence avoiding to add the extra package there). I'm all for adding it in the main package.
I guess I can drop the env variable there too, as I was setting it up so that the same script would work for all versions, but if we add support everywhere, we're good :) 

> There is also some strange indentation, so please make sure that it is consistent across the entire file.

Oops, you're right. I've done that from a fresh installed server before updating my vim config, and it tried putting tabs everywhere even though the indentation clearly wasn't. I'll fix that up.

> I have not worked with this type of challenge before, so I would like to do some more research about how we best implement this in a sustainable way in case we want to add more DNS providers.

I don't know many of the plugins for DNS01 challenges. I know that Cloudflare is one, and that you can also connect to your local bind for it to refresh your local DNS server automatically. That latter one might be more complicated to implement with containers. That being said, we could support any plugins by using "dns-cloudflare" (the plugin name) in the cert name, for instance (+ the parameter for default plugin to use could be that too). I'll make those changes and update the PR!

XaF commented:
Ok, updated with a more generic approach.

From [the certbot documentation](), we can find the different authenticators using DNS-01:
```
  --dns-cloudflare      Obtain certificates using a DNS TXT record (if you are
                        using Cloudflare for DNS). (default: False)
  --dns-cloudxns        Obtain certificates using a DNS TXT record (if you are
                        using CloudXNS for DNS). (default: False)
  --dns-digitalocean    Obtain certificates using a DNS TXT record (if you are
                        using DigitalOcean for DNS). (default: False)
  --dns-dnsimple        Obtain certificates using a DNS TXT record (if you are
                        using DNSimple for DNS). (default: False)
  --dns-dnsmadeeasy     Obtain certificates using a DNS TXT record (if you are
                        using DNS Made Easy for DNS). (default: False)
  --dns-gehirn          Obtain certificates using a DNS TXT record (if you are
                        using Gehirn Infrastructure Service for DNS).
                        (default: False)
  --dns-google          Obtain certificates using a DNS TXT record (if you are
                        using Google Cloud DNS). (default: False)
  --dns-linode          Obtain certificates using a DNS TXT record (if you are
                        using Linode for DNS). (default: False)
  --dns-luadns          Obtain certificates using a DNS TXT record (if you are
                        using LuaDNS for DNS). (default: False)
  --dns-nsone           Obtain certificates using a DNS TXT record (if you are
                        using NS1 for DNS). (default: False)
  --dns-ovh             Obtain certificates using a DNS TXT record (if you are
                        using OVH for DNS). (default: False)
  --dns-rfc2136         Obtain certificates using a DNS TXT record (if you are
                        using BIND for DNS). (default: False)
  --dns-route53         Obtain certificates using a DNS TXT record (if you are
                        using Route53 for DNS). (default: False)
  --dns-sakuracloud     Obtain certificates using a DNS TXT record (if you are
                        using Sakura Cloud for DNS). (default: False)
```

The question for you @JonasAlfredsson - do we want to add support for all of those ?
While I'm here, I don't mind adding support for all of them (except for `rfc2136` since it's not an external provider, but a bind server).

XaF commented:
Last commit adds support for all DNS01 providers listed above, except for `rfc2136`.

JonasAlfredsson commented on src/scripts/run_certbot.sh:
```diff
@@ -49,12 +49,39 @@ fi
+    local challenge_params=
+    local challenge_type=
+
+    if [ -n "$challenge_authenticator" ] && [[ "$challenge_authenticator" == dns-* ]]; then
```
Could probably be shortened to 

```
if [[ "${4}" == dns-* ]]; then 
    ...
else 
   challenge_authenticator="webroot" ...
fi
```

since we have made sure that there will always be something in `$4` from the caller.

XaF replied:
I actually shortened the variable names a bit, but kept it in `$authenticator` since it's more talking than `$4`, and does not require to rewrite anything here since we're failing when we don't know the authenticator now!

XaF replied:
Well, I actually went ahead and replaced it by `$4`

JonasAlfredsson commented on src/scripts/run_certbot.sh:
```diff
@@ -87,6 +114,14 @@ for conf_file in /etc/nginx/conf.d/*.conf*; do
+            challenge_authenticator=${BASH_REMATCH[2]}
+            debug "Found mention of authenticator '${challenge_authenticator}' in name '${cert_name}'"
+        else
+            challenge_authenticator="${CERTBOT_DEFAULT_AUTHENTICATOR:-webroot}"
```
I would like to squeeze in something like this here as well

```
if [ -n "${CERTBOT_DEFAULT_AUTHENTICATOR}" ] || [[ "${CERTBOT_DEFAULT_AUTHENTICATOR,,}" == webroot ]]; then 
     challenge_authenticator="webroot"
else 
    error "Unknown authenticator '${CERTBOT_DEFAULT_AUTHENTICATOR}'"
    exit 1
fi
```

XaF replied:
Do we want to `exit 1` or skip to next cert ?

Also, slightly confused by the logic here?

- `if DEFAULT_AUTHENTICATOR is not empty || DEFAULT_AUTHENTICATOR is webroot`
   Redundant, did you mean -z instead of -n ?

- `else error unknown authenticator DEFAULT_AUTHENTICATOR`
   Do you want to check that value against the authenticators list ?

XaF replied:
So I kept the logic simple here, but made the full check in `get_certificates`, where we actually check that the authenticator we got as parameter exists!

JonasAlfredsson commented on src/scripts/run_certbot.sh:
```diff
@@ -49,12 +49,39 @@ fi
+        if [ -f "$configfile" ]; then
+            challenge_type="dns-01"
+            challenge_params="--${challenge_authenticator}-credentials=${configfile}"
+            local challenge_envvar="CERTBOT_DNS_${provider^^}_PROPAGATION_SECONDS"
```
This line can be removed if you change to the more generic name suggested above.

JonasAlfredsson commented on src/scripts/run_certbot.sh:
```diff
@@ -49,12 +49,39 @@ fi
+                challenge_params="${challenge_params} --${challenge_authenticator}-propagation-seconds=${!challenge_envvar}"
+            fi
+        else
+            debug "Authenticator is ${challenge_authenticator} but ${configfile} is missing; defaulting to webroot"
```
This should be an error, where we `return 1` directly after.

XaF replied:
So we prefer failing entirely than trying webroot, makes sense! :)

JonasAlfredsson reviewed (commented):
This is a really good change, I like it and thank you for taking the time to make it.
However, there are some places where some small changes can really improve the flow and readability of the code :)

There is only one concern that I have, and it is described in the very long comment. 
As I have said earlier, I have not worked with DNS challenges before, so is there any benefit of doing it in case you are not interested in a wildcard certificate?

XaF commented on README.md:
```diff
@@ -70,6 +70,8 @@ instructions, from `@staticfloat`'s image, can be found
 - `ELLIPTIC_CURVE`: The size/[curve][15] of the ECDSA keys (default: `secp256r1`)
 - `USE_ECDSA`: Set to `1` to have certbot use [ECDSA instead of RSA](./docs/good_to_know.md#ecdsa-and-rsa-certificates) (default: `0`)
 - `RENEWAL_INTERVAL`: Time interval between certbot's [renewal checks](./docs/good_to_know.md#renewal-check-interval) (default: `8d`)
+- `CERTBOT_DEFAULT_AUTHENTICATOR`: The default authenticator plugin to use when not defined directly in the certificate path (default: `webroot`)
```
Makes sense! Will replace :)

XaF commented on README.md:
```diff
@@ -70,6 +70,8 @@ instructions, from `@staticfloat`'s image, can be found
 - `USE_ECDSA`: Set to `1` to have certbot use [ECDSA instead of RSA](./docs/good_to_know.md#ecdsa-and-rsa-certificates) (default: `0`)
 - `RENEWAL_INTERVAL`: Time interval between certbot's [renewal checks](./docs/good_to_know.md#renewal-check-interval) (default: `8d`)
+- `CERTBOT_DEFAULT_AUTHENTICATOR`: The default authenticator plugin to use when not defined directly in the certificate path (default: `webroot`)
+- `CERTBOT_<plugin>_PROPAGATION_SECONDS` (e.g. `CERTBOT_DNS_CLOUDFLARE_PROPAGATION_SECONDS`): The number of seconds to wait after a DNS challenge has been setup before asking Let's Encrypt to check it (defaults to certbot's default)
```
Hesitated between how much precise we wanted to be if using different providers at the same time, but I guess you're right: we don't need to be that precise!

XaF commented on src/Dockerfile:
```diff
@@ -28,6 +43,8 @@ RUN set -ex && \
 # Install certbot.
     pip3 install -U cffi certbot \
+# And the supported extra authenticators
+        $(echo $CERTBOT_DNS_AUTHENTICATORS | sed 's/\(^\| \)/\1certbot-dns-/g') \
```
It does! I actually pushed to the docker hub for some tests: https://hub.docker.com/r/thexaf/nginx-certbot (if you wanna try it/play with it)

XaF commented on src/scripts/run_certbot.sh:
```diff
@@ -87,6 +114,14 @@ for conf_file in /etc/nginx/conf.d/*.conf*; do
+
         # Find all 'server_names' in this .conf file and assemble the list of
         # domains to be included in the request.
         domain_request=""
```
DNS challenges _allow_ to support wildcard domains, but do not necessarily mean that you want a certificate for a wildcard domain. One of the big interests of DNS challenges is that you can get a certificate even if the traffic is not directly pointing at you at the moment (allowing for instance to have a local server with valid certificates that's not reachable from outside). At the moment, I'm rewriting my server's configuration to be in containers, and I want the setup fully working before migrating where the traffic points to: DNS challenges allow me to do that, webroot wouldn't. And if I want to support a part of my domains through wildcards, I can! But I don't _have to_ (and most of my domains won't have wildcard handling).

Let's Encrypt will by itself complain if you have a wildcard and a conflicting domain with it (e.g. `*.example.com` and `first.example.com` would conflict together in the cert) but will allow non-conflicting domains together (e.g. `*.example.com` and `*.first.example.com` will work, since the wildcard is supposed to cover only a single level). I believe the domains logic should be kept as it is right now, and people should fix their configuration accordingly when using wildcards. If you declare your server name entry with both `*.example.com` and `first.example.com` for instance, the latter is redundant, so that'd be good practice to just keep the wildcard anyway, I think!

XaF commented on src/Dockerfile:
```diff
@@ -3,6 +3,22 @@ LABEL maintainer="Jonas Alfredsson <jonas.alfredsson@protonmail.com>"
+    luadns \
+    nsone \
+    ovh \
+    rfc2136 \
```
I actually saw that this plugin is configured the same as others, and thus someone could have their bind in a different container and be able to work with it. So I added it here too!

JonasAlfredsson replied:
Sounds good!

XaF commented:
@JonasAlfredsson any other changes needed for this to be mergeable ? :)

JonasAlfredsson commented on src/scripts/run_certbot.sh:
```diff
@@ -49,12 +49,39 @@ fi
+    local challenge_type=
+
+    if [[ "$4" == dns-* ]] && \
+            [[ ${CERTBOT_DNS_AUTHENTICATORS} =~ (^| )${4#dns-}( |$) ]]; then
```
Nit: keep this on the same line as above, it will look more clean.

JonasAlfredsson commented on src/scripts/run_certbot.sh:
```diff
@@ -49,12 +49,39 @@ fi
+    local authenticator_params=
+    local challenge_type=
+
+    if [[ "$4" == dns-* ]] && \
```
Just a small thing here, should we have "${4,,}" in all of these places for extra safety?

XaF replied:
Adding it, hesitated of doing it where we're calling the function, but it was done in the function for other parameters, so doing the same! :)

JonasAlfredsson commented on src/scripts/run_certbot.sh:
```diff
@@ -49,12 +49,39 @@ fi
+        challenge_type="http-01"
+        authenticator_params="--webroot-path=/var/www/letsencrypt"
+    else
+        error "Unknown authenticator ${4} for '${1}'"
```
Put single quotes around this one`'${4}'`

JonasAlfredsson commented on src/scripts/run_certbot.sh:
```diff
@@ -49,12 +49,39 @@ fi
+        local provider="${4#dns-}"
+        local configfile="/etc/letsencrypt/${provider}.ini"
+        if [ ! -f "$configfile" ]; then
+            error "Authenticator is ${4} but ${configfile} is missing"
```
I would like single quotes around the variables so it will be easier to identify them when reading the logs later.

JonasAlfredsson reviewed (commented):
This is starting to look really good. 
We should expand the documentation to reflect all of these new features, but that can be done in the future. 
Just take a look at the last final touches, and then I think this will be possible to merge :)

JonasAlfredsson commented:
The docker images are building correctly, and the code looks like it would be backwards compatible without users doing anything which means that we should probably be able to get away with just a bump to the minor version number. 
I have not had time to test the actual functionality out myself, so I trust that you have tried it out on your end? :)

XaF commented:
Added documentation, and addressed your last comments! 

As per your question, I'm currently already using my PR-ed container for my server, until we can merge this and get the new version ready so I can revert to using yours ;) - It works well!

```
Found mention of authenticator 'dns-cloudflare' in name '<REDACTED>-dns-cloudflare'
Requesting an RSA certificate for '<REDACTED>-dns-cloudflare' (dns-01 through dns-cloudflare)
Requesting a certificate for <REDACTED>

Successfully received certificate.
Certificate is saved at: /etc/letsencrypt/live/<REDACTED>-dns-cloudflare/fullchain.pem
Key is saved at:         /etc/letsencrypt/live/<REDACTED>-dns-cloudflare/privkey.pem
This certificate expires on 2022-02-09.
These files will be updated when the certificate renews.
NEXT STEPS:
- The certificate will need to be renewed before it expires. Certbot can automatically renew the certificate in the background, but you may need to take steps to enable that functionality. See https://certbot.org/renewal-setup for instructions.
Saving debug log to /var/log/letsencrypt/letsencrypt.log

- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
If you like Certbot, please consider supporting our work by:
 * Donating to ISRG / Let's Encrypt:   https://letsencrypt.org/donate
 * Donating to EFF:                    https://eff.org/donate-le
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
Found mention of authenticator 'dns-cloudflare' in name '<REDACTED>-dns-cloudflare'
Requesting an RSA certificate for '<REDACTED>-dns-cloudflare' (dns-01 through dns-cloudflare)
Certificate not yet due for renewal

- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
Certificate not yet due for renewal; no action taken.
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
Saving debug log to /var/log/letsencrypt/letsencrypt.log
```

XaF commented:
As another note: I reverted back to use the `local authenticator` variable in the `get_certificates` method, as I needed an intermediate variable to lowercase, while still being able to trim the `dns-` prefix to get the provider name.

Alternatively, I could be lowercasing when we're calling the `get_certificates` method, but that would be weird because we're not doing any of the other lowercasing there...

JonasAlfredsson commented on src/scripts/run_certbot.sh:
```diff
@@ -87,6 +114,14 @@ for conf_file in /etc/nginx/conf.d/*.conf*; do
         fi
 
+        # Determine the authenticator to use to solve the challenge
+        if [[ "${cert_name,,}" =~ (^|[^a-z0-9])(dns-($(echo ${CERTBOT_DNS_AUTHENTICATORS} | sed 's/ /|/g')))([^a-z0-9]|$) ]]; then
```
I am not sure changing to `(^|[^a-z0-9])` and `([^a-z0-9]|$)` is such a good idea, I think we should keep the delimiter consistent between this and the `key_type` lookup and `^a-z0-9` allows way too many options in my opinion :)

XaF replied:
You're right, I can keep the same delimiters, this is too broad.
However, I'll keep the `(^|[.-])` and  `([.-]|$)` if you're good with that, as they allow for delimited-only (either by one of the delimiters or the beginning/end of the string) matching, vs `dns-cloudflarey` matching because it starts by `dns-cloudflare`. What do you think ?
I believe we could apply the same matching to RCC/RSA also, as it's backward compatible too.

JonasAlfredsson replied:
That is really clean, I like it a lot!
Sure, this should be copied to the key matching, but let's do that after this is merged :)

JonasAlfredsson commented on docs/certbot_authenticators.md:
````diff
@@ -0,0 +1,92 @@
+```
+server {
+    listen              443 ssl;
+    server_name         yourdomain.org *.yourdomain.org;
````
Question here, do both we and Nginx support having a wildcard in the server name?

XaF replied:
So for the container aspect, yes we do:
```
Found mention of authenticator 'dns-cloudflare' in name 'wildcard.<REDACTED>-dns-cloudflare'
Requesting an RSA certificate for 'wildcard.<REDACTED>-dns-cloudflare' (dns-01 through dns-cloudflare)
Requesting a certificate for *.<REDACTED> and <REDACTED>

Successfully received certificate.
Certificate is saved at: /etc/letsencrypt/live/wildcard.<REDACTED>-dns-cloudflare/fullchain.pem
Key is saved at:         /etc/letsencrypt/live/wildcard.<REDACTED>-dns-cloudflare/privkey.pem
This certificate expires on 2022-02-15.
These files will be updated when the certificate renews.
NEXT STEPS:
- The certificate will need to be renewed before it expires. Certbot can automatically renew the certificate in the background, but you may need to take steps to enable that functionality. See https://certbot.org/renewal-setup for instructions.
Saving debug log to /var/log/letsencrypt/letsencrypt.log

- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
If you like Certbot, please consider supporting our work by:
 * Donating to ISRG / Let's Encrypt:   https://letsencrypt.org/donate
 * Donating to EFF:                    https://eff.org/donate-le
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
Found all the necessary files for '/etc/nginx/conf.d/30-wildcard.conf.nokey', enabling...
```

And as per nginx:
```
$ curl -vvv --connect-to ::127.0.0.1 https://test.<REDACTED>/
* Connecting to hostname: 127.0.0.1
*   Trying 127.0.0.1:443...
* Connected to 127.0.0.1 (127.0.0.1) port 443 (#0)
* ALPN, offering h2
* ALPN, offering http/1.1
* successfully set certificate verify locations:
*  CAfile: /etc/ssl/certs/ca-certificates.crt
*  CApath: /etc/ssl/certs
* TLSv1.3 (OUT), TLS handshake, Client hello (1):
* TLSv1.3 (IN), TLS handshake, Server hello (2):
* TLSv1.3 (IN), TLS handshake, Encrypted Extensions (8):
* TLSv1.3 (IN), TLS handshake, Certificate (11):
* TLSv1.3 (IN), TLS handshake, CERT verify (15):
* TLSv1.3 (IN), TLS handshake, Finished (20):
* TLSv1.3 (OUT), TLS change cipher, Change cipher spec (1):
* TLSv1.3 (OUT), TLS handshake, Finished (20):
* SSL connection using TLSv1.3 / TLS_AES_256_GCM_SHA384
* ALPN, server accepted to use http/1.1
* Server certificate:
*  subject: CN=*.<REDACTED>
*  start date: Nov 17 19:13:32 2021 GMT
*  expire date: Feb 15 19:13:31 2022 GMT
*  subjectAltName: host "test.<REDACTED>" matched cert's "*.<REDACTED>"
*  issuer: C=US; O=Let's Encrypt; CN=R3
*  SSL certificate verify ok.
> GET / HTTP/1.1
> Host: test.<REDACTED>
> User-Agent: curl/7.74.0
> Accept: */*
>
* TLSv1.3 (IN), TLS handshake, Newsession Ticket (4):
* TLSv1.3 (IN), TLS handshake, Newsession Ticket (4):
* old SSL session ID is stale, removing
* Mark bundle as not supporting multiuse
< HTTP/1.1 200 OK
< Server: nginx
< Date: Wed, 17 Nov 2021 20:22:00 GMT
< Content-Type: text/html
< Content-Length: 10
< Last-Modified: Wed, 17 Nov 2021 20:19:58 GMT
< Connection: keep-alive
< ETag: "619563ee-a"
< Accept-Ranges: bytes
<
It works!
* Connection #0 to host 127.0.0.1 left intact
```

JonasAlfredsson replied:
Oh, I did not know that Nginx could take [wildcards in the server name](http://nginx.org/en/docs/http/server_names.html).
Yeas, you have indeed proven that this will work ;)

JonasAlfredsson commented on docs/certbot_authenticators.md:
````diff
@@ -0,0 +1,92 @@
+}
+```
+
+The script ran in the container to renew certificates will automatically identify that it needs to use the Route53 authenticator here. Of course, you will need that authenticator to be configured properly in order to be able to use it.
````
Add a note here stating that it will work with the RSA/ECC configuration at the same time? 
https://github.com/JonasAlfredsson/docker-nginx-certbot/blob/master/docs/advanced_usage.md#multi-certificate-setup

JonasAlfredsson reviewed (commented):
Once again excellent work!
Very nice to hear that you are able to get it to work on your end :)
The only concern I have now is the changed regex, I think it should at least be consistent with how the RSA/ECC matching works.

JonasAlfredsson commented:
I don't really have more comments here and now, so unless you want to add something extra I don't see why this couldn't be merged :)

XaF commented:
> I don't really have more comments here and now, so unless you want to add something extra I don't see why this couldn't be merged :)

I'm fine with that, too!! 🎉 
Thanks for the back and forth to make this better, @JonasAlfredsson !

JonasAlfredsson commented:
This is now merged, and will be built for the `:latest` tag. 
I will probably try to add some small extra things before I make a new release version though.

maxkagamine commented:
Just wanted to say thanks, both of you! This was just what I needed; worked like a charm, wildcard cert too 👍

JonasAlfredsson commented:
Oh my, that was a fast adoption of a new feature. Really nice to hear :)
Just because I have not yet had time to work with wildcard domains, would you mind sharing how your Nginx config is set up so I can get a better understanding on how/when this may come to use?

maxkagamine commented:
I'm setting up some docker containers on my NAS and needed DNS validation since they'll be mostly internal; my thinking was I'd use a single wildcard cert as they'll all be subdomains. (Your timing here was impeccable btw; when I found the repo looking for a simple container with nginx+letsencrypt, your comment this was ready to be merged was just a few hours ago!)

Looking at the code, though, it doesn't seem there's a way to have sites share a cert. Not a huge deal since with the way your script automatically parses out sites' domain names, I frankly didn't even need a wildcard (I'm also using it for the Unraid web ui with [a small script](https://gist.github.com/maxkagamine/5b6c34db6045d6413db3b333d6d2bae2) to copy the cert from the container, but in retrospect I could have just explicitly added its domain). 

That said it might be a nice improvement, e.g. aggregate cert name→domains before calling `get_certificate`. I could give it a try later and send a PR. Unless, of course, this totally already works and I've simply overlooked something, which is entirely possible :)

JonasAlfredsson commented:
Alright, glad to hear that this was such a nice coincidence :)

But yes, I have also been thinking about how to best handle sharing of a single wildcard certificate over multiple configs/sites.
A solution (which starts to make this entrypoint script complicated enough to make me want to move to a another language) would perhaps to create an [associative array](https://www.artificialworlds.net/blog/2012/10/17/bash-associative-array-examples/) and store `[${cert_name}] = "domain_name,domain_name"` and then iterate over that in the end. 
I don't really like it, because it looks like the value here would really benefit from being a list and Bash does not support that. However, there are workarounds to that (using `cut -d,`), so I think we should be fine.

The more difficult part is how to define you want this to be a wildcard certificate. Perhaps we can add something like this:

```
ssl_certificate_key     /etc/letsencrypt/live/mydomain.org/privkey.pem; #wildcard
```

where we then pick up the wildcard comment and make this request become `[mydomain.org] = "mydomain.org,*.mydomain.org"`. So in this case the `cert_name` must be the base domain? 

This discussion should probably continue in the two new issues I have created:
- https://github.com/JonasAlfredsson/docker-nginx-certbot/issues/91
- https://github.com/JonasAlfredsson/docker-nginx-certbot/issues/92
